### PR TITLE
[AppSec] Benchmark

### PR DESCRIPTION
**Pull request by ValentinZakharov**

### CODE REVIEW

File: dd-java-agent/appsec/appsec.gradle [hunk 1]
```diff
@@ -41,6 +45,15 @@ jar {
   archiveClassifier = 'unbundled'
 }
 
+jmh {
+  fork = 5
+  warmupForks = 0
+  jmhVersion = '1.28'
```
Comment (richardstartin) [inline dd-java-agent/appsec/appsec.gradle:51]: 1.32 is the latest

File: dd-java-agent/appsec/appsec.gradle [hunk 2]
```diff
@@ -41,6 +45,15 @@ jar {
...
+  duplicateClassesStrategy = DuplicatesStrategy.EXCLUDE
+  jvmArgsAppend = ['-Ddd.appsec.beta=true']
```
Comment (richardstartin) [inline dd-java-agent/appsec/appsec.gradle:53]: I like to keep heap size constant and pretouch here, and explicitly set a GC

File: dd-java-agent/appsec/appsec.gradle [hunk 3]
```diff
@@ -16,6 +17,9 @@ dependencies {
   implementation project(':internal-api')
   implementation group: 'io.sqreen', name: 'libsqreen', version: '2.0'
 
+  jmh group: 'org.openjdk.jmh', name: 'jmh-generator-annprocess', version: '1.28'
+  testAnnotationProcessor 'org.openjdk.jmh:jmh-generator-annprocess:1.28'
+
```
Comment (bantonsson) [inline dd-java-agent/appsec/appsec.gradle:22]: I have no idea why this would be necessary

File: dd-java-agent/appsec/appsec.gradle [hunk 4]
```diff
@@ -41,6 +45,16 @@ jar {
...
 }
 
+jmh {
+  fork = 5
+  warmupForks = 0
+  jmhVersion = '1.32'
+  duplicateClassesStrategy = DuplicatesStrategy.EXCLUDE
+  jvmArgsAppend = ['-Ddd.appsec.beta=true -Xms64m -Xmx64m']
+  failOnError = false
+  forceGC = false
```
Comment (bantonsson) [inline dd-java-agent/appsec/appsec.gradle:55]: I think you should force a GC.

File: dd-java-agent/appsec/src/jmh/java/datadog/appsec/benchmark/AppSecBenchmark.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import static java.util.concurrent.TimeUnit.*;
+
+@State(Scope.Benchmark)
+@Warmup(iterations = 5, time = 1, timeUnit = SECONDS)
+@Measurement(iterations = 5, time = 1, timeUnit = SECONDS)
```
Comment (bantonsson) [inline dd-java-agent/appsec/src/jmh/java/datadog/appsec/benchmark/AppSecBenchmark.java:18]: Having `5` runs of `1s` each seems very short for the JVM to be able to optimize anything. But OTOH ~60% of the time in the benchmark seems to be spent inside the native code from the profiling information.

File: dd-java-agent/appsec/appsec.gradle (hunk 3 above)
Comment (ValentinZakharov) [reply dd-java-agent/appsec/appsec.gradle]: Oh, that I was experimenting with debugging jmh. (removed)

File: dd-java-agent/appsec/appsec.gradle (hunk 4 above)
Comment (ValentinZakharov) [reply dd-java-agent/appsec/appsec.gradle]: Thanks - copy/paste by mistake.
Fixed

File: dd-java-agent/appsec/appsec.gradle (hunk 2 above)
Comment (ValentinZakharov) [reply dd-java-agent/appsec/appsec.gradle]: I enabled force GC and define fixed heap size.

File: dd-java-agent/appsec/appsec.gradle (hunk 1 above)
Comment (ValentinZakharov) [reply dd-java-agent/appsec/appsec.gradle]: Thanks, version - updated!

File: dd-java-agent/appsec/src/jmh/java/datadog/appsec/benchmark/AppSecBenchmark.java (hunk 5 above)
Comment (ValentinZakharov) [reply dd-java-agent/appsec/src/jmh/java/datadog/appsec/benchmark/AppSecBenchmark.java:18]: Do you have any thoughts about the best timing could be the best to use?

Comment (bantonsson) [reply dd-java-agent/appsec/src/jmh/java/datadog/appsec/benchmark/AppSecBenchmark.java:17]: This is fine for now.

**Review by bantonsson — APPROVED** [review]

**Pull request merged**